src/ctapipe/image/extractor.py

```diff
@@ -561,26 +576,34 @@ def __call__(
     ) -> DL1CameraContainer:
         if self.pixel_fraction.tel[tel_id] == 1.0:
             # average over pixels then argmax over samples
-            peak_index = waveforms[~broken_pixels].mean(axis=-2).argmax()
+            peak_index = waveforms[:, ~broken_pixels].mean(axis=-2).argmax(axis=-1)
         else:
             n_pixels = int(self.pixel_fraction.tel[tel_id] * waveforms.shape[-2])
-            brightest = np.argsort(waveforms.max(axis=-1))[~broken_pixels][
+            brightest = np.argsort(waveforms.max(axis=-1))[:, ~broken_pixels][
                 ..., -n_pixels:
             ]
 
             # average over brightest pixels then argmax over samples
-            peak_index = waveforms[brightest].mean(axis=-2).argmax()
+            peak_index = (
+                waveforms[:, brightest][:, 0, ...].mean(axis=-2).argmax(axis=-1)
+            )
 
         charge, peak_time = extract_around_peak(
             waveforms,
-            peak_index,
+            peak_index[:, np.newaxis],
             self.window_width.tel[tel_id],
             self.window_shift.tel[tel_id],
             self.sampling_rate_ghz[tel_id],
         )
         if self.apply_integration_correction.tel[tel_id]:
-            charge *= self._calculate_correction(tel_id=tel_id)[selected_gain_channel]
-        return DL1CameraContainer(image=charge, peak_time=peak_time, is_valid=True)
+            correction = self._calculate_correction(tel_id=tel_id)
+            if selected_gain_channel is None:
+                charge *= correction[:, np.newaxis]
+            else:
+                charge *= correction[selected_gain_channel]
+        return DL1CameraContainer(
+            image=np.squeeze(charge), peak_time=np.squeeze(peak_time), is_valid=True
+        )
```

Comment:
_Sonarqube_ is complaining that this block is duplicated verbatim in about 4 places. Could we maybe move it into the parent class? Or maybe refactor it into a common a helper function?  That would reduce potential errors if somebody fixes a bug later.